### ISSUE TITLE
联邦学习时loss问题

### ISSUE THREAD

**Issue by StarvenLan** [SMILELab-FL/FedLab#186]
问题描述：

   情况一：train与val都放在本地。
   情况二：train在本地，val在服务器，采用联邦学习策略。

情况二中val的loss远远大于情况一中val的loss，请问是什么问题呢？

**Comment by Zengdun-cs** [comment]
如果你使用的是fedlab.utils.functional中的evaluate函数的话，FedLab这里的实现给出的loss的计算结果是sum而不是avg。
因此，loss大小受验证数据集大小的影响。

**Comment by StarvenLan** [comment]
这里我使用的是自己的loss计算，不是fedlab.utils.functional中的evaluate函数。

**Comment by AgentDS** [comment]
没理解错的话你是想比较centerized training和FL training的性能差异是吗？

**Comment by StarvenLan** [comment]
不算是，我现在是想先跑fedavg，然后看看它的性能如何，比如loss与precision，后面改了联邦学习的算法后进行比较，但是目前情况是联邦学习的val的loss远远超过本地val的loss（数据集完全一样，这里我对比测试时联邦学习只有一个client），我觉得这里面不太合理，所以提出了这个疑问。

**Comment by Zengdun-cs** [comment]
可以贴一下你的相关代码吗

**Comment by StarvenLan** [comment]
> service:
        # FedLab 
        serialized_params = SerializationTool.serialize_model(model_train)
        print(" weight_serialized_params:", serialized_params, "    size:", serialized_params.shape)

        Fedlogger = Logger(log_name="server")
        handler = SyncParameterServerHandler(model_train,
                                             global_round=opt.round,
                                             logger=Fedlogger,
                                             sample_ratio=opt.sample)

        manager_ = ServerSynchronousManager(handler=handler,
                                            network=network,
                                            logger=Fedlogger)
        manager_.setup()
        t_start = time_sync()
        with open(os.path.join(loss_history.save_path, "time.txt"), 'a') as f:
            f.write("comm_round:  now_val_time/s:     wait_time/s:    from_start_to_now/mins:   receive_time_point:")
            f.write("\n")
        manager_.main_loop_yolo(gen_val, Cuda, optimizer, yolo_loss, loss_history, epoch_step_val, tb_writer, opt)
        manager_.shutdown()
        t_end = time_sync()
        print("all_time:",t_end - t_start)
        -------------------------------------------------------------------------------------------------------------------------------------
    def main_loop_yolo(self, gen_val, Cuda, optimizer, yolo_loss, loss_history, epoch_step_val, tb_writer, opt):
        comm_round = 0
        while self._handler.stop_condition() is not True:
            print("communicate_round:",comm_round+1)
 
            activate = threading.Thread(target=self.activate_clients)
            activate.start()
            while True:
                sender, message_code, payload = PackageProcessor.recv_package()
                if message_code == MessageCode.ParameterUpdate:
                    model_parameters = payload[0]
                    if self._handler.add_model(sender, model_parameters):
                        break
                else:
                    raise Exception(
                        "Unexpected message code {}".format(message_code))

            val_loss = 0
            model_train = self._handler.get_model()

            model_train.eval()
            print('Start Validation')
            for iteration, batch in enumerate(tqdm(gen_val)):
                images, targets = batch[0], batch[1]
                with torch.no_grad():
                    if Cuda:
                        images = torch.from_numpy(images).type(torch.FloatTensor).cuda()
                        targets = [torch.from_numpy(ann).type(torch.FloatTensor).cuda() for ann in targets]
                    else:
                        images = torch.from_numpy(images).type(torch.FloatTensor)
                        targets = [torch.from_numpy(ann).type(torch.FloatTensor) for ann in targets]

                    optimizer.zero_grad()
                    outputs = model_train(images)

                    loss_value_all = 0
                    num_pos_all = 0
                    for l in range(len(outputs)):
                        loss_item, num_pos = yolo_loss(l, outputs[l], targets)
                        loss_value_all += loss_item
                        num_pos_all += num_pos
                    loss_value = loss_value_all / num_pos_all

                val_loss += loss_value.item()

            print('Finish Validation')
            print(' Val Loss: %.3f ' % (val_loss))

            Recall, Precision = getmap(model_train)
            comm_round = comm_round + 1
        -------------------------------------------------------------------------------------------------------------------------------------
         def get_model(self):
                return self.model



> client:
    # FedLab 
    criterion = nn.CrossEntropyLoss()
    Fedlogger = Logger(log_name="client " + str(opt.rank))
    trainer = ClientSGDTrainer(
        model_train,
        gen,
        epochs=1,
        optimizer=optimizer,
        criterion=criterion,
        cuda=device.type != 'cpu',
        logger=Fedlogger
    )
    manager_ = ClientPassiveManager(trainer=trainer,
                                    network=network,
                                    logger=Fedlogger)
    manager_.setup()
    comm_round = 0

    while True:
        sender_rank, message_code, payload = PackageProcessor.recv_package(src=0)
        if message_code == MessageCode.Exit:
            break
        elif message_code == MessageCode.ParameterUpdate:
            model_parameters = payload[0]
            print("shape:", model_parameters.shape)

            SerializationTool.deserialize_model(model_train, model_parameters)  # load parameters
            print("Local train procedure is running")

        print("communicate_round:",comm_round + 1)


        for epoch in range(start_epoch, end_epoch):
            loss = 0
            val_loss = 0

            model_train.train()
            print('Start Train')
            with tqdm(total=epoch_step, desc=f'Epoch {epoch + 1}/{end_epoch}', postfix=dict, mininterval=0.3) as pbar:
                for iteration, batch in enumerate(gen):
                    if iteration >= epoch_step:
                        break

                    images, targets = batch[0], batch[1]
                    with torch.no_grad():
                        if Cuda:
                            images = torch.from_numpy(images).type(torch.FloatTensor).cuda()
                            targets = [torch.from_numpy(ann).type(torch.FloatTensor).cuda() for ann in targets]
                        else:
                            images = torch.from_numpy(images).type(torch.FloatTensor)
                            targets = [torch.from_numpy(ann).type(torch.FloatTensor) for ann in targets]

                    optimizer.zero_grad()
                    outputs = model_train(images)

                    loss_value_all = 0
                    num_pos_all = 0

                    for l in range(len(outputs)):
                        loss_item, num_pos = yolo_loss(l, outputs[l], targets)
                        loss_value_all += loss_item
                        num_pos_all += num_pos
                    loss_value = loss_value_all / num_pos_all

                    loss_value.backward()
                    optimizer.step()

                    loss += loss_value.item()

                    pbar.set_postfix(**{'loss': loss / (iteration + 1),
                                        'lr': get_lr(optimizer)})
                    pbar.update(1)

            print('Finish Train')
            print('Total Loss: %.3f' % (loss / epoch_step))


        comm_round = comm_round + 1

        sender_rank = SerializationTool.serialize_model(model_train)
        pack = Package(message_code=MessageCode.ParameterUpdate,
                       content=sender_rank)
        print(
            "----------------------------------------------------------------------------------------------------")
        print("pack:", pack)
        PackageProcessor.send_package(pack, dst=0)
    else:
        raise ValueError("Invalid MessageCode {}. Please see MessageCode Enum".format(message_code))

**Comment by Zengdun-cs** [comment]
是对同一个模型，在server上的验证表现和client上的验证表现的loss不同？
我觉得你得把你的问题说清楚些。
如果是FedAvg和集中学习的损失问题，那可能是算法原因。
如果是逻辑上应该是一样的模型，但是实际测试时表现又不一样，有可能是实现的差别，或者有参数传输时的精度损失。

**Comment by StarvenLan** [comment]
是集中学习时val的loss与联邦学习val的loss差别很大的问题。

**Comment by Zengdun-cs** [comment]
那这个问题我们无法给你解答，这个问题不是FedLab的bug。 
关于研究的问题上，FedAvg算法不一定适用于所有深度学习模型的联邦优化，受影响的因素很多，这应该是你要去研究的问题。

**Comment by AgentDS** [comment]
> 是集中学习时val的loss与联邦学习val的loss差别很大的问题。

单个client上数据集怎么样？如果单个client上包含完整training set，那么相同round*epoch下应该跟集中训练是一样的，model aggregate的时候就只有一个model参与。如果单个client上数据只是集中训练用到的数据集的子集，那确实会有性能差异

**Comment by AgentDS** [comment]
只有一个client的FedAvg原理上和集中训练一样啊，只是数据集不知道你怎么分布的

**Comment by StarvenLan** [comment]
是单个client上包含完整training set的情况，所以这个结果我才感觉有点困惑。

**Comment by Zengdun-cs** [comment]
可以测试一下，client和server间的模型参数传输是否有错误。
